### DISCUSSION TITLE
Integration rtx with launchctl / macOS GUI apps

### DISCUSSION THREAD

**Discussion by eproxus** [jdx/mise#68] [General]
I have perhaps a peculiar use-case (although it shouldn't be too uncommon):

I start GUI apps like Sublime Text, that needs access to rtx-managed versions for e.g. LSP functionality. When using asdf I could statically add `~/.asdf/shims` and `~/.asdf/bin` to the launchctl path which would make GUI apps just use whatever was globally activated with asdf. 

However, with rtx there is no such static path. Setting the launchctl path can only be done manually and after which the Dock (!) in macOS has to be restarted (or you have to log out and in again). Obviously this is not realistic to use, because as soon as you've upgraded your globally configure version of a tool the old configured path would be out of date.

What would be a practical way to solve this with rtx? Maintain some kind of symlink on shell startup, like the global `~/.asdf/shims`?

**Comment by eproxus** [comment]
Would it be too much to ask of rtx to symlink path like `~/.local/share/rtx/installs/<lang>/global/bin` whenever the global version is changed?

**Reply by jdx** [reply]
this isn't necessary, we can create shims that don't need to know the version since they call `rtx exec`. `rtx exec` will automatically use the correct version of the runtime based on `$PWD`.

**Comment by jdx** [comment]
I'm opposed to using shims in the general case. They solve some problems like this, but they come with a ton of baggage. At some point though it may make sense to add the ability for rtx to generate shims and I think it could work for some alternative use-cases like this.

This isn't high on my priority though, but you could write your own shims really easily:

```
bash-5.2$ ~/.rtx/shims/node -v
v18.13.0
bash-5.2$ mkdir -p ~/.rtx/shims
bash-5.2$ cat >~/.rtx/shims/node <<EOF
#!/usr/bin/env bash
exec rtx exec -- node "\$@"
EOF
bash-5.2$ chmod +x ~/.rtx/shims/node
bash-5.2$ ~/.rtx/shims/node -v
v18.13.0
```

You'll need to do this for each binary you want to support though.

Alternatively, we are doing basically the exact same thing as direnv, so you could also fork the direnv extension for whatever IDE you're using and make it work for RTX: https://github.com/misuzu/direnv-subl

**Reply by eproxus** [reply]
Funny you should mention direnv and direnv-subl 😄 I just installed that before you answered and that solved my problem exactly!

**Reply by eproxus** [reply]
And for what it's worth, I agree about the shims. It was just the only solution I could think of at the moment

**Reply by jdx** [reply]
how does the direnv plugin solve your problem? are you using the `use rtx` hook?

**Reply by eproxus** [reply]
Yes, exactly. I have rtx installed, I have direnv installed (via Homebrew, as to not have a bootstrapping problem. I need Sublime to be able to access direnv, and besides I don't need multiple direnv versions installed). Then I have the following in my `~/.envrc`:

```sh
use rtx
```

Then direnv-subl uses that which in turn activates rtx inside somehow. Funnily enough, running `print(os.environ['PATH'])` inside Sublime _doesn't_ show the correct path, but Sublime-LSP picks it up somehow (as can be seen when running the action _LSP: Troubleshoot Server_ which lists the `PATH` used) 🤷

**Reply by eproxus** [reply]
Spoke too soon. Apparently the direnv plugin didn't help (I guess it doesn't understand `use rtx`?). I solved it by adding shims instead (😢).

The following is a universal shim that can be symlinked to any executable name and will automatically work:
```sh
#!/bin/sh

exec rtx exec -- "$(basename $0)" "$@"
```

I put it in `~/.local/bin` (which is in my path) as `rtx_shim` and then symlink anything I need (e.g. `ln -s ~/.local/bin/{rtx_shim,node}`).

**Reply by jdx** [reply]
look at the rtx readme for setting up the `use rtx` direnv extension

**Reply by jdx** [reply]
btw your shim needs to look like this:

```
exec rtx exec -- "$(basename $0)" "$@"
```

the quotes on the first probably won't matter (add them anyways for consistency though), but they are required on the second if you have any arguments with spaces in them

**Reply by eproxus** [reply]
> look at the rtx readme for setting up the `use rtx` direnv extension

I did follow all instructions related to direnv. But the shims fit my workflow much more, since I want to continue to use RTX in the shell and not replace everything with direnv.

**Reply by jdx** [reply]
oh I think I understand, did you mean that you couldn't use the direnv sublime plugin because it didn't work with the extension?

**Reply by eproxus** [reply]
I guess so. I have it enabled, I run LSP and can run Troubleshoot Server but don't see an updated path there, unfortunately (no idea if it is supposed to work, could also be something with the LSP plug-in I suppose).

**Comment by jdx** [comment]
btw if you control which binary it uses, say you're calling launchctl and you want it to use the right python, then just use `rtx exec -- python` instead of `python`.

**Reply by jdx** [reply]
this just isn't likely to work for something like a LSP in an IDE because you can't actually change this

**Comment by amoosbr** [comment]
**Update (2023-03-07):** Since the introduction of experimental shim support, I use them for potential IDE needs. The script mentioned below is no longer used.

I created a small helper tool for me to create symlinked PATHS for my global tool binaries.
More or less the following lines for every global tool:
```sh
global_symlink_dir="$HOME/opt/global-path-symlinks"
# Python
python_base="$global_symlink_dir/python"
mkdir -p "$python_base"
ln -snf "$(rtx where python)/bin" "$python_base/bin"
```

In the PATH, I can then use: `/Users/user/opt/global-path-symlinks/python/bin:`

In general, it is sufficient for me, that apps just have this global version available.
The script is quite simple and not fully automated, but sufficient for my usage.  It's available in this [gist](https://gist.github.com/amoosbr/de91b39541f5cfde7a8003d2eb6fd9d9), if someone wants to look at it.

With this solution, I only need to update the launchctl path, when a tool is added or removed from my global tool list.

**Reply by jdx** [reply]
@amoosbr I'm curious what you think of https://github.com/jdxcode/rtx/issues/92. I think the current behavior actually works well for your particular use-case. (`rtx which python` will either use the version specified in `.tool-versions` OR the latest installed if none specified).
